### ISSUE TITLE
SHM-2051: Task: eth_newPendingTransactionFilter

### ISSUE THREAD

**Issue by chrischabot** [shardeum-mirror/shardeum#36]
field | value
| - | - |
url | https://linear.app/shm/issue/SHM-2051/task-eth-newpendingtransactionfilter
author | @
date | Fri Aug  4 05:44:51 BST 2023
state | Done
project | RPC: Implement remaining RPC methods
priority | High
assignee | @
labels | 
related | SHM-2052
parent | 
children | 
PRs | 
attachments | 

1. There was also one that we moved to a future project because it has some tech dependencies that are not ready just yet: eth_newPendingTransactionFilter
2. \*\**July 10, 2023 1:29 PM (EDT)July 10, 2023 1:29 PM (EDT)July 10, 2023 1:29 PM (EDT)*

   (I think we need a new task for eth_newPendingTransactionFilter)

   **Andrew:**

   I am working on a design for either the explorer or archiver to be able to send a stream of transactions to a third party microserverice. I think we will implement a microservice that provides eth_newPendingTransactionFilter that is co hosted with the json rpc server and possible cohosted with an archiver or explorer collector

   **Andrew**:
   1. I think we make move the eth_newPendingTransactionFilter out of the stateful project so that the other work you did can get considered tested and done, and then eth_newPendingTransactionFilter can wait on the additional tech
   2. \*\**July 10, 2023 10:54 AM (GMT+6:30)July 10, 2023 10:54 AM (GMT+6:30)July 10, 2023 10:54 AM (GMT+6:30)*

      this is better than keeping a mostly done project floating around
      
      [https://www.quicknode.com/docs/ethereum/eth_newPendingTransactionFilter](https://www.quicknode.com/docs/ethereum/eth_newPendingTransactionFilter)

**Comment by chrischabot** [comment]
field | value
|-|-|
url | https://linear.app/shm/issue/SHM-2051/task-eth-newpendingtransactionfilter#comment-e9d74697
author | @
date | Fri Aug  4 05:45:07 BST 2023

![image.png](https://uploads.linear.app/1065affa-aebb-4031-b7e5-3fb32c4187d8/cc6b8a36-6de1-42ef-bf24-8b8efc4f8921/89cdaf69-6ae6-4723-be15-f4e3bc6f9eb5)

**Comment by chrischabot** [comment]
field | value
|-|-|
url | https://linear.app/shm/issue/SHM-2051/task-eth-newpendingtransactionfilter#comment-f69efb20
author | @
date | Wed Sep 27 22:28:44 BST 2023

* [X] I'm not sure,  but this method seems to not be implemented yet or hardcoded on purpose, but will need to ask Thanht, Joe, or Ahmed. 

  Original implementation

  ```
    eth_newPendingTransactionFilter: async function (args: any, callback: any) {
      const api_name = 'eth_newPendingTransactionFilter'
      const ticket = crypto
        .createHash('sha1')
        .update(api_name + Math.random() + Date.now())
        .digest('hex')
      logEventEmitter.emit('fn_start', ticket, api_name, performance.now())
      if (verbose) {
        console.log('Running newPendingTransactionFilter', args)
      }
      const result = '0x1'
      callback(null, result)
      logEventEmitter.emit('fn_end', ticket, { success: true }, performance.now())
    },
  ```

**Comment by chrischabot** [comment]
field | value
|-|-|
url | https://linear.app/shm/issue/SHM-2051/task-eth-newpendingtransactionfilter#comment-cc8b57c5
author | @
date | Thu Sep 28 15:01:03 BST 2023

~So instead of changing anything since it seems like this might be done on purpose. I will set a few pending transaction and try to request the filter ID with ~[~eth_newPendingTransactionFilter~](https://www.quicknode.com/docs/ethereum/eth_newPendingTransactionFilter)~. Then use the filter ID with ~`~eth_getFilterChanges~`~ to get data associated with the subscription.~

**Comment by chrischabot** [comment]
field | value
|-|-|
url | https://linear.app/shm/issue/SHM-2051/task-eth-newpendingtransactionfilter#comment-2aef4cb2
author | @
date | Thu Sep 28 16:46:16 BST 2023

~No resulting pending actions. Laptop didn't like this~


`curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"eth_newPendingTransactionFilter","params":[],"id":1}' `[`http://localhost:8080`](http://localhost:8080)

`curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"eth_getFilterChanges","params":["0x1"],"id":1}' `[`http://localhost:8080`](http://localhost:8080)

`npx hardhat load_test --type eth_transfer --tps 100 --duration 60 --eoa 10 --validate true`

![image.png](https://uploads.linear.app/1065affa-aebb-4031-b7e5-3fb32c4187d8/da8389ff-e6b1-4b73-83a9-afaa6f0280a1/294edf4b-7d0d-4e21-964a-839dd5d062ae)

![image.png](https://uploads.linear.app/1065affa-aebb-4031-b7e5-3fb32c4187d8/219acff0-c2dc-4f7f-940e-61831e6cb10f/925d96d4-7e35-4c3a-b990-c248220af915)

**Comment by chrischabot** [comment]
field | value
|-|-|
url | https://linear.app/shm/issue/SHM-2051/task-eth-newpendingtransactionfilter#comment-1d6751ca
author | @
date | Thu Sep 28 23:38:19 BST 2023

Need to do unit testing for each of our methods

**Comment by chrischabot** [comment]
field | value
|-|-|
url | https://linear.app/shm/issue/SHM-2051/task-eth-newpendingtransactionfilter#comment-e7c5a259
author | @
date | Fri Sep 29 14:46:09 BST 2023

From Thant:

I think `eth_newPendingTransactionFilter` is very similar to `eth_newFilter` and `eth_getFilterChanges`. You should first review how we handle `newFilter` and `getFilterChanges` API. The `filterId` is effectively like a subscription ID to a certain event. For your case, when RPC receives `eth_newPendingTransactionFilter` request, 

* it should create a subscription ID (or filterID), store the ID in the memory or database, and return the filterID to the client.
* \[\] Then, after some time the client will call `eth_getFilterChanges` by providing the `filterId` to see if there are new pending transactions.
* The RPC server should use that `filterId`, load the filter data (for instance contract address, block number) from memory or database and *query the pending transactions from the explorer*, and return to the client.

**Comment by chrischabot** [comment]
field | value
|-|-|
url | https://linear.app/shm/issue/SHM-2051/task-eth-newpendingtransactionfilter#comment-2f08d40c
author | @
date | Fri Sep 29 16:21:59 BST 2023

* [ ] ~check out eth_newFilter & eth_getFilterChanges
  Thant answers main question that it seems like filterId should not be hardcoded~
  * [X] ~Need to understand how to filter data from memory/database and query pending txns from the explorer. (out of scope for this task?)~
  * [X] ~look into subscription~

**Comment by chrischabot** [comment]
field | value
|-|-|
url | https://linear.app/shm/issue/SHM-2051/task-eth-newpendingtransactionfilter#comment-2fb55371
author | @
date | Wed Oct  4 18:07:24 BST 2023

~Need to test out~

* [X] ~getPendingTransactions() by using~ 
* [X] ~eth_getFilterChanges and~
* [X] ~eth_newPendingTransactionFilter~ 



Testing
shardeum-server

1. enter `export NODE_ENV='debug'`


2. enter `shardus start 15`

json-rpc-server

1. `npm` `run prepare` or `compile`
2. `npm run start`

load-tester

1. to have pending transactions enter `npx hardhat load_test --type eth_transfer --tps 100 --duration 60 --eoa 10 --validate true`
2. to make reqeuest to `eth_newPendingTransactionFilter`. 
   1. `curl -X POST `[`http://localhost:8080`](http://localhost:8080)` -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","id":1,"method":"eth_newPendingTransactionFilter","params":[]}'`
3. to make a request to `eth_getFilterChanges`
   1. `curl -X POST --data '{"jsonrpc":"2.0","method":"eth_getFilterChanges","params":["0x1"],"id":1}' `[`http://localhost:8080`](http://localhost:8545)

**Comment by chrischabot** [comment]
field | value
|-|-|
url | https://linear.app/shm/issue/SHM-2051/task-eth-newpendingtransactionfilter#comment-3a187516
author | @
date | Wed Oct  4 22:17:54 BST 2023

* [X] not sure if correct filterObj is being saved or if more info should be added. Not sure if extra should be added based on references
* [X] Should there only be one filter object mapped for just pending transactions and that means when `eth_newPendingTransactionFilter` one object is created or do we want a unique object for each? Would it matter if a bunch of filter objects for pending transactions are in `filtersMap`. Seems like other filter objects that are called by `eth_getFilterChanges` are updated if a filterID has been called before checking for new changes since the last call. 
  * [X] I guess not because of this example from alchamy? [https://docs.alchemy.com/reference/eth-newpendingtransactionfilter-polygon#:\~:text=eth_newPendingTransactionFilter%20,state%20has%20changed%2C%20call%20eth_getFilterChanges](https://docs.alchemy.com/reference/eth-newpendingtransactionfilter-polygon#:\~:text=eth_newPendingTransactionFilter%20,state%20has%20changed%2C%20call%20eth_getFilterChanges)
* [X] question from me: so createdBlock is saved to the filterObj… When retrieving pending transactions it's getting the transactions that has not been processed in the last 15 seconds and not really using any of the data in the filterObj…. Not really sure how createdBlock can be used since I don't understand how abstract blocks are created so I need to ask this.
